test/integration/pulp_upload_test.go

```diff
@@ -0,0 +1,187 @@
+
+	router := echo.New()
+	router.HTTPErrorHandler = config.CustomHTTPErrorHandler
+	router.Use(middleware.WrapMiddlewareWithSkipper(identity.EnforceIdentity, middleware.SkipAuth))
+	pathPrefix := router.Group(api.FullRootPath())
+
+	handler.RegisterPulpRoutes(pathPrefix, s.dao)
+
+	s.server = &http.Server{
+		Addr:    "127.0.0.1:8000",
+		Handler: router,
+	}
+
+	// start the server in a goroutine
+	go func() {
+		if err := s.server.ListenAndServe(); err != nil {
+			log.Fatal().Msg("Could not start server")
+		}
+	}()
+
+	// force local storage for integration tests
+	config.Get().Clients.Pulp.StorageType = "local"
+}
+
+func (s *UploadSuite) servePulpRouter(req *http.Request) (int, []byte, error) {
+	rr := httptest.NewRecorder()
+	s.server.Handler.ServeHTTP(rr, req)
+
+	response := rr.Result()
+	defer response.Body.Close()
+
+	body, err := io.ReadAll(response.Body)
+	require.NoError(s.T(), err)
+
+	return response.StatusCode, body, err
+}
+
+func TestUploadSuite(t *testing.T) {
+	suite.Run(t, new(UploadSuite))
+}
+
+func (s *UploadSuite) TestUpload() {
+	t := s.T()
+
+	// Create an upload
+	createRequest := api.CreateUploadRequest{
+		Size: 4,
+	}
+	var uploadResponse zest.UploadResponse
+
+	body, err := json.Marshal(createRequest)
+	require.NoError(t, err)
+
+	path := api.FullRootPath() + "/pulp/uploads/"
+	req := httptest.NewRequest(http.MethodPost, path, bytes.NewReader(body))
+	req.Header.Set(api.IdentityHeader, test_handler.EncodedIdentity(t))
+	req.Header.Set("Content-Type", "application/json")
+
+	code, body, err := s.servePulpRouter(req)
+	assert.Nil(t, err)
+	assert.Equal(t, http.StatusOK, code)
+	err = json.Unmarshal(body, &uploadResponse)
+	assert.Nil(t, err)
+	assert.Contains(t, string(body), "pulp_href")
+
+	// Upload a file chunk
+
+	// add multipart request
+	fileBytes := new(bytes.Buffer)
+	multipartWriter := multipart.NewWriter(fileBytes)
+
+	// create sha256 hasher
+	hasher := sha256.New()
+
+	// add form field for file and write file content to hasher
+	filePart, err := multipartWriter.CreateFormFile("file", "test-rpm-chunk")
+	assert.Nil(t, err)
+	fileContent := []byte(randomFileContent(int(createRequest.Size)))
+	_, err = filePart.Write(fileContent)
+	assert.Nil(t, err)
+	hasher.Write(fileContent)
+	err = multipartWriter.Close()
+	assert.Nil(t, err)
+
+	// calculate checksum of the data written to the hasher
+	uploadSha256 := hex.EncodeToString(hasher.Sum(nil))
+
+	path = api.FullRootPath() + "/pulp/uploads/" + *uploadResponse.PulpHref
+	req = httptest.NewRequest(http.MethodPut, path, fileBytes)
+	req.Header.Set(api.IdentityHeader, test_handler.EncodedIdentity(t))
+	req.Header.Set("Content-Type", multipartWriter.FormDataContentType())
+	req.Header.Set("Content-Range", fmt.Sprintf("bytes 0-%d/*", createRequest.Size-1))
+
+	code, body, err = s.servePulpRouter(req)
+	assert.Nil(t, err)
+	assert.Equal(t, http.StatusOK, code)
+	assert.Contains(t, string(body), "pulp_href")
+	fmt.Println(string(body))
```

Comment:
should probably switch this to a debug log, or remove it :)

Reply:
lol oops! will remove